File: demo_src/CPP/3D/Sine_surface/sine_surf.cpp
#include <cstdio>
#include <iostream> 
#include <iomanip> 
#include <cstdlib>
#include <cmath>
#include "sine_surf.h"

typedef const double creal;
typedef const int cint;
typedef double real;
using namespace std;

//* -------------------------------------------------------------------------- *
//* DESCRIPTION (reference phase where f(x,y,z) < 0):                          *
//* sinusoidal surface inside the cube [0,1]x[0,1]x[0,1]                       *
//* f(x,y,z) = z - a0 - b0*sin(c1*pi*x + pi*d1)*sin(c1*pi*x + pi*e1)           *
//* -------------------------------------------------------------------------- *

static double a0 = 0.5;
static double b0 = 1./6.;
static double c1 = 1.6;
static double d1 = 1./7.;
static double e1 = 1./5.;


/* -------------------------------------------------------------------------- */

void init()
{

  /* initialize random number generator. */
  srand(time(0));  

  double scalingfactor = 0.05;
 
  /* b0 --> from 1.15 to 1.20 */
  b0 = 1.15 + ((double)rand() / RAND_MAX)*scalingfactor;
  
  /* c1 --> from 1.6 to 1.65 */
  c1 = 1.6 + ((double)rand() / RAND_MAX)*scalingfactor;
   
  /* d1 --> from 0.12 to 0.17 */
  d1 = 0.12 + ((double)rand() / RAND_MAX)*scalingfactor;
 
  /* e1 --> from 0.15 to 0.25 */
  e1 = 0.15 + ((double)rand() / RAND_MAX)*scalingfactor;
  
  return;
}


real impl_func(creal xy[])
{
  double x,y,z,f0;

  x = xy[0];
  y = xy[1];
  z = xy[2];

  f0 = z - a0 - b0*sin(M_PI*(c1*x+d1))*sin(M_PI*(c1*y+e1));

  return f0;
}

//* -------------------------------------------------------------------------- *

void check_volume(creal vol_n)
{
  double vol_a;

  //vol_a = 0.5;
  vol_a = a0 + (b0/(c1*M_PI*c1*M_PI))*(cos(d1*M_PI) - cos((d1+c1)*M_PI))*(cos(e1*M_PI) - cos((e1+c1)*M_PI)); 

  cout << "-----------------------------------------------------------" << endl;
  cout << "--------------- CPP: sinusoidal surface check -------------" << endl;
  cout << "analytical volume: " << scientific << setw(23) << setprecision(16) 
       << vol_a << endl;
  cout << "numerical  volume: " << scientific << setw(23) << setprecision(16) 
       << vol_n << endl << endl;
  cout << "absolute error   : " << scientific << setw(23) << setprecision(16) 
       << fabs(vol_a-vol_n) << endl;
  cout << "relative error   : " << scientific << setw(23) << setprecision(16) 
       << fabs(vol_a-vol_n)/vol_a << endl;
  cout << "-----------------------------------------------------------" << endl;
//   cout << "with Intel i7 3.4 GHz + Linux openSUSE 13.1 + gcc 4.8.1 -O2" << endl;
//   cout << "-----------------------------------------------------------" << endl;
//   cout << "analytical volume:  5.0000000000000000e-01" << endl;	     
//   cout << "numerical  volume:  5.0000000000000011e-01" << endl << endl; 
//   cout << "absolute error   :  1.1102230246251565e-16" << endl;	     
//   cout << "relative error   :  2.2204460492503131e-16" << endl;         
//   cout << "------------- CPP: end sinusoidal surface check -----------" << endl;
//   cout << "-----------------------------------------------------------" << endl 
  cout << endl;
  
  return;
}

File: demo_src/Fortran/3D/Sine_surface/sine_surf.f90
!* -------------------------------------------------------------------------- *
!* DESCRIPTION (reference phase where f(x,y,z) < 0):                          *
!* sinusoidal surface inside the cube [0,1]x[0,1]x[0,1]                       *
!* f(x,y,z) = z - A0 - B0*sin(C1*pi*x + pi*D1)*sin(C1*pi*x + pi*E1)           *
!* -------------------------------------------------------------------------- *

MODULE SINESURFACE_MOD

  IMPLICIT NONE

  REAL(8), PARAMETER ::  MYPI = 3.141592653589793238462643D0
  REAL(8) ::  A0 = 0.5D0, B0 = (1.D0/6.D0), C1 = 1.6D0
  REAL(8) ::  D1 = (1.D0/7.D0), E1 = (1.D0/5.D0)

  CONTAINS

!  initialize a random seed from the system clock at every run 
SUBROUTINE INIT_RANDOM_SEED()

      INTEGER :: i, n, clock
      INTEGER, DIMENSION(:), ALLOCATABLE :: seed
      INTRINSIC RANDOM_SEED, SYSTEM_CLOCK 

      CALL RANDOM_SEED(size = n)
      ALLOCATE(seed(n))

      CALL SYSTEM_CLOCK(COUNT=clock)

      seed = clock + 37 * (/ (i - 1, i = 1, n) /)
      CALL RANDOM_SEED(PUT = seed)

      DEALLOCATE(seed)
      
END SUBROUTINE INIT_RANDOM_SEED


SUBROUTINE INIT()  

  IMPLICIT NONE
  
  REAL(8) :: seed
  REAL(8), PARAMETER :: scalingfactor  = 0.05D0
    
  INTRINSIC RANDOM_NUMBER  
  
  CALL INIT_RANDOM_SEED()
  
  !B0 --> from 1.15 to 1.20 
  CALL RANDOM_NUMBER(seed) 
  B0 = 1.15D0 + seed*scalingfactor
  
  !C1 --> from 1.6 to 1.65
  CALL RANDOM_NUMBER(seed) 
  C1 = 1.6D0 + seed*scalingfactor
   
  !D1 --> from 0.12 to 0.17
  CALL RANDOM_NUMBER(seed) 
  D1 = 0.12D0 + seed*scalingfactor
 
  !E1 --> from 0.15 to 0.25
  CALL RANDOM_NUMBER(seed) 
  E1 = 0.15D0 + seed*scalingfactor
  
END SUBROUTINE INIT

!* -------------------------------------------------------------------------- *
  
REAL(8) FUNCTION IMPL_FUNC (xyz)

  IMPLICIT NONE
  REAL(8), DIMENSION(3), INTENT(IN) :: xyz
  REAL(8) :: x,y,z

  INTRINSIC DSIN

  x = xyz(1)
  y = xyz(2)
  z = xyz(3)
  
  IMPL_FUNC = z - A0 - B0*DSIN(MYPI*(C1*x+D1))*DSIN(MYPI*(C1*y+E1))

  RETURN

END FUNCTION IMPL_FUNC

!* -------------------------------------------------------------------------- *

SUBROUTINE CHECK_VOLUME(volnum)

  IMPLICIT NONE
  REAL(8),INTENT(IN) :: volnum
  REAL(8) :: volana

  INTRINSIC DABS, DCOS

  !volana = 0.5D0;
  volana = A0 + (B0/(C1*MYPI*C1*MYPI))*(DCOS(D1*MYPI) - DCOS((D1+C1)*MYPI))*(DCOS(E1*MYPI) - DCOS((E1+C1)*MYPI)); 

  write(*,*) '-----------------------------------------------------------'
  write(*,*) '--------------- F: sinusoidal surface check ---------------'
  write(*,100) volana
  write(*,101) volnum
  write(*,*) ' '
  write(*,102) DABS(volnum-volana)
  write(*,103) DABS(volnum-volana)/volana
  write(*,*) '-----------------------------------------------------------'
!   write(*,*) 'with Intel i7 3.4 GHz + Linux openSUSE 13.1 + gcc 4.8.1 -O2'
!   write(*,*) '-----------------------------------------------------------'
!   write(*,*) 'analytical volume:  5.0000000000000000E-01'
!   write(*,*) 'numerical  volume:  4.9999999999999994E-01'
!   write(*,*) ' '
!   write(*,*) 'absolute error   :  5.5511151231257827E-17'
!   write(*,*) 'relative error   :  1.1102230246251565E-16'
!   write(*,*) '------------- F: end sinusoidal surface check -------------'
!   write(*,*) '-----------------------------------------------------------'
  write(*,*) ' '
   100 FORMAT(' analytical volume: ', ES23.16)
   101 FORMAT(' numerical  volume: ', ES23.16)
   102 FORMAT(' absolute error   : ', ES23.16)
   103 FORMAT(' relative error   : ', ES23.16)

END SUBROUTINE CHECK_VOLUME

END MODULE SINESURFACE_MOD